### PR TITLE
various optimisations/fixes

### PR DESCRIPTION
**Pull request by jimpil**
Hi there, 

We are very likely to use this at work, so I just went ahead and tidied-up slightly, removed reflective calls completely, fixed a bug (see `read-dates-as-instants!`) etc etc. The thing I really wanted was to be able to avoid lazy-seqs in queries, so now that is possible with a new `:realise-fn` option (where applicable). It defaults to `sequence`, so nothing changes for users, but now, one can also pass `(partial into [])`. 

Let me know what you think :+1:

### CODE REVIEW

File: src/mongo_driver_3/model.clj [hunk 1]
```diff
@@ -16,8 +16,8 @@
 
 (defn read-dates-as-instants! []
   (extend-protocol ConvertToDocument
-    Date 
-    (from-document [input _]
+    Date
+    (document [input _]
```
Comment (jimpil) [inline src/mongo_driver_3/model.clj:20]: help me understand please...should this protocol extension be `ConvertToDocument`/`document`, or `ConvertFromDocument`/`from-document`? It **certainly** cannot be  `ConvertToDocument`/`from-document`, right?

**Comment by gnarroway** [comment]
Thanks for this. Will have a look soon

**Comment by gnarroway** [comment]
merged thanks. in future (esp re the realise-fn), would be nice to have tests, adjust readme, and write a commit msg that explains it (the what and the why).

**Comment by jimpil** [comment]
> merged thanks. in future (esp re the realise-fn), would be nice to have tests, adjust readme, and write a commit msg that explains it (the what and the why).

Understood - i just wasn't sure if you were going to look at these changes in a positive light. I am/was open to iterating on the PR.

About the protocol extension mismatch (see my very first message) - it looks like a serious bug...are you going to address it for the next release? Many thanks in advance...

**Pull request merged**